pkg/expr/convert_to_long.go

```diff
@@ -0,0 +1,314 @@
+			if !field.Type().Numeric() {
+				continue
+			}
+			newField := data.NewFieldFromFieldType(field.Type(), field.Len())
+			newField.Name = field.Name
+			newField.Labels = field.Labels.Copy()
+			for i := 0; i < field.Len(); i++ {
+				newField.Set(i, field.CopyAt(i))
+			}
+			newFrame.Fields = append(newFrame.Fields, newField)
+		}
+	}
+	return data.Frames{newFrame}
+}
+
+func convertNumericWideToNumericLong(frames data.Frames) (data.Frames, error) {
+	// Wide should only be one frame
+	if len(frames) != 1 {
+		return nil, fmt.Errorf("expected exactly one frame for wide format, but got %d", len(frames))
+	}
+	inputFrame := frames[0]
+
+	// The Frame should have no more than one row
+	if inputFrame.Rows() > 1 {
+		return nil, fmt.Errorf("expected no more than one row in the frame, but got %d", inputFrame.Rows())
+	}
+
+	// Gather Unique Label Keys from each numeric, and unique numeric field names,
+	// each one maps to a field in the output long Frame.
+	uniqueNames := make([]string, 0)
+	uniqueKeys := make([]string, 0)
+
+	uniqueNamesMap := make(map[string]data.FieldType)
+	uniqueKeysMap := make(map[string]struct{})
+
+	prints := make(map[string]int)
+
+	registerPrint := func(labels data.Labels) {
+		fp := labels.Fingerprint().String()
+		if _, ok := prints[fp]; !ok {
+			prints[fp] = len(prints)
+		}
+	}
+
+	for _, field := range inputFrame.Fields {
+		if field.Type().Numeric() {
+			if _, ok := uniqueNamesMap[field.Name]; !ok {
+				uniqueNames = append(uniqueNames, field.Name)
+			}
+			uniqueNamesMap[field.Name] = field.Type()
+
+			if field.Labels != nil {
+				for key := range field.Labels {
+					if _, ok := uniqueKeysMap[key]; !ok {
+						uniqueKeys = append(uniqueKeys, key)
+					}
+					uniqueKeysMap[key] = struct{}{}
+					registerPrint(field.Labels)
+				}
+			}
+		}
+	}
+
+	// Create new fields for output Long frame
+	fields := make([]*data.Field, 0, len(uniqueNames)+len(uniqueKeys))
+
+	// Create the Numeric Fields, tracking the index of each field by name
+	// Note: May want to use FloatAt and and prepopulate with NaN so missing
+	// combinations of value can be NA instead of the zero value of 0.
+	var nameIndexMap = make(map[string]int, len(uniqueNames))
+	for i, name := range uniqueNames {
+		field := data.NewFieldFromFieldType(uniqueNamesMap[name], len(prints))
+		field.Name = name
+		fields = append(fields, field)
+		nameIndexMap[name] = i
+	}
+
+	// Create the String field, tracking the index of each field by key
+	var keyIndexMap = make(map[string]int, len(uniqueKeys))
+	for i, k := range uniqueKeys {
+		fields = append(fields, data.NewField(k, nil, make([]string, len(prints))))
+		keyIndexMap[k] = len(nameIndexMap) + i
+	}
+
+	longFrame := data.NewFrame("", fields...)
+
+	if inputFrame.Rows() == 0 {
+		return data.Frames{longFrame}, nil
+	}
+
+	// Add Rows to the fields
+	for _, field := range inputFrame.Fields {
+		if !field.Type().Numeric() {
+			continue
+		}
+		rowIdx := prints[field.Labels.Fingerprint().String()]
```

Comment:
Not sure, but this feels more like a column index than a row-index to me. At least considering the source, rather than the destination. Just my gut feel I guess.
```suggestion
		columnIdx := prints[field.Labels.Fingerprint().String()]
```